File: bindings/python/iota_sdk/types/feature.py
# Copyright 2023 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

from enum import IntEnum
from typing import Dict, List, TypeAlias, Union, Any
from dataclasses import dataclass, field
from dataclasses_json import config
from iota_sdk.types.address import Address, deserialize_address
from iota_sdk.types.block_issuer_key import BlockIssuerKey
from iota_sdk.types.common import EpochIndex, HexStr, hex_str_decoder, json, SlotIndex


class FeatureType(IntEnum):
    """Types of features.

    Attributes:
        Sender (0): The sender feature.
        Issuer (1): The issuer feature.
        Metadata (2): The metadata feature.
        StateMetadata (3): The state metadata feature.
        Tag (4): The tag feature.
        NativeToken (5): The native token feature.
        BlockIssuer (6): The block issuer feature.
        Staking (7): The staking feature.
    """
    Sender = 0
    Issuer = 1
    Metadata = 2
    StateMetadata = 3
    Tag = 4
    NativeToken = 5
    BlockIssuer = 6
    Staking = 7


@json
@dataclass
class SenderFeature:
    """Identifies the validated sender of an output.
    Attributes:
        address: A given sender address.
    """
    type: int = field(
        default_factory=lambda: int(
            FeatureType.Sender),
        init=False)
    address: Address = field(
        metadata=config(
            decoder=deserialize_address
        ))


@json
@dataclass
class IssuerFeature:
    """Identifies the validated issuer of the UTXO state machine.
    Attributes:
        address: A given issuer address.
    """
    type: int = field(
        default_factory=lambda: int(
            FeatureType.Issuer),
        init=False)
    address: Address = field(
        metadata=config(
            decoder=deserialize_address
        ))


@json
@dataclass
class MetadataFeature:
    """Defines metadata, arbitrary binary data, that will be stored in the output.
    Attributes:
        entries: A key-value map where the keys are graphic ASCII strings and the values hex-encoded binary data.
    """
    type: int = field(
        default_factory=lambda: int(
            FeatureType.Metadata),
        init=False)
    entries: Dict[str, HexStr]


@json
@dataclass
class StateMetadataFeature:
    """A Metadata Feature that can only be changed by the State Controller.
    Attributes:
        entries: A key-value map where the keys are graphic ASCII strings and the values hex-encoded binary data.
    """
    type: int = field(
        default_factory=lambda: int(
            FeatureType.StateMetadata),
        init=False)
    entries: Dict[str, HexStr]


@json
@dataclass
class TagFeature:
    """Makes it possible to tag outputs with an index, so they can be retrieved through an indexer API.
    Attributes:
        tag: A hex encoded tag used to index the output.
    """
    type: int = field(default_factory=lambda: int(FeatureType.Tag), init=False)
    tag: HexStr


@json
@dataclass
class NativeTokenFeature:
    """Contains a native token.
        id: The unique identifier of the native token.
        amount: The amount of native tokens.
    """
    type: int = field(default_factory=lambda: int(
        FeatureType.NativeToken), init=False)
    id: HexStr
    amount: int = field(metadata=config(
        encoder=hex,
        decoder=hex_str_decoder,
    ))


@json
@dataclass
class BlockIssuerFeature:
    """Contains the public keys to verify block signatures and allows for unbonding the issuer deposit.
    Attributes:
        expiry_slot: The slot index at which the Block Issuer Feature expires and can be removed.
        block_issuer_keys: The Block Issuer Keys.
    """
    type: int = field(
        default_factory=lambda: int(
            FeatureType.BlockIssuer),
        init=False)
    expiry_slot: SlotIndex
    block_issuer_keys: List[BlockIssuerKey]


@json
@dataclass
class StakingFeature:
    """Stakes IOTA coins to become eligible for committee selection, validate the network and receive Mana rewards.
    Attributes:
        staked_amount: The amount of IOTA coins that are locked and staked in the containing account.
        fixed_cost: The fixed cost of the validator, which it receives as part of its Mana rewards.
        start_epoch: The epoch index in which the staking started.
        end_epoch: The epoch index in which the staking ends.
    """
    type: int = field(
        default_factory=lambda: int(
            FeatureType.Staking),
        init=False)
    staked_amount: int = field(metadata=config(
        encoder=str
    ))
    fixed_cost: int = field(metadata=config(
        encoder=str
    ))
    start_epoch: EpochIndex
    end_epoch: EpochIndex


Feature: TypeAlias = Union[SenderFeature, IssuerFeature,
                           MetadataFeature, StateMetadataFeature, TagFeature, NativeTokenFeature, BlockIssuerFeature, StakingFeature]


# pylint: disable=too-many-return-statements
def deserialize_feature(d: Dict[str, Any]) -> Feature:
    """
    Takes a dictionary as input and returns an instance of a specific class based on the value of the 'type' key in the dictionary.

    Arguments:
    * `d`: A dictionary that is expected to have a key called 'type' which specifies the type of the returned value.
    """
    feature_type = d['type']
    if feature_type == FeatureType.Sender:
        return SenderFeature.from_dict(d)
    if feature_type == FeatureType.Issuer:
        return IssuerFeature.from_dict(d)
    if feature_type == FeatureType.Metadata:
        return MetadataFeature.from_dict(d)
    if feature_type == FeatureType.StateMetadata:
        return StateMetadataFeature.from_dict(d)
    if feature_type == FeatureType.Tag:
        return TagFeature.from_dict(d)
    if feature_type == FeatureType.NativeToken:
        return NativeTokenFeature.from_dict(d)
    if feature_type == FeatureType.BlockIssuer:
        return BlockIssuerFeature.from_dict(d)
    if feature_type == FeatureType.Staking:
        return StakingFeature.from_dict(d)
    raise Exception(f'invalid feature type: {feature_type}')


def deserialize_features(dicts: List[Dict[str, Any]]) -> List[Feature]:
    """
    Takes a list of dictionaries as input and returns a list with specific instances of classes based on the value of the 'type' key in the dictionary.

    Arguments:
    * `dicts`: A list of dictionaries that are expected to have a key called 'type' which specifies the type of the returned value.
    """
    return list(map(deserialize_feature, dicts))

File: bindings/nodejs/lib/types/block/output/feature.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { SlotIndex } from '../slot';
import { Transform, Type } from 'class-transformer';
import { Address, AddressDiscriminator } from '../address';
import {
    BlockIssuerKey,
    BlockIssuerKeyDiscriminator,
} from './block-issuer-key';
import { u256, u64 } from '../../utils/type-aliases';
import { EpochIndex } from '../../block/slot';
import { NativeToken } from '../../models/native-token';
import { HexEncodedString, hexToBigInt } from '../../utils/hex-encoding';

/**
 * Printable ASCII characters.
 */
export declare type PrintableASCII = string;

/**
 * All of the feature block types.
 */
enum FeatureType {
    /** A Sender feature. */
    Sender = 0,
    /** An Issuer feature. */
    Issuer = 1,
    /** A Metadata feature. */
    Metadata = 2,
    /** A StateMetadata feature. */
    StateMetadata = 3,
    /** A Tag feature. */
    Tag = 4,
    /** A NativeToken feature. */
    NativeToken = 5,
    /** A BlockIssuer feature. */
    BlockIssuer = 6,
    /** A Staking feature. */
    Staking = 7,
}

/** The base class for features. */
abstract class Feature {
    readonly type: FeatureType;

    /**
     * @param type The type of feature.
     */
    constructor(type: FeatureType) {
        this.type = type;
    }
}

/**
 * A Sender feature.
 */
class SenderFeature extends Feature {
    @Type(() => Address, {
        discriminator: AddressDiscriminator,
    })
    readonly address: Address;

    /**
     * @param sender The Sender address stored with the feature.
     */
    constructor(sender: Address) {
        super(FeatureType.Sender);
        this.address = sender;
    }
}

/**
 * An Issuer feature.
 */
class IssuerFeature extends Feature {
    @Type(() => Address, {
        discriminator: AddressDiscriminator,
    })
    readonly address: Address;

    /**
     * @param issuer The Issuer address stored with the feature.
     */
    constructor(issuer: Address) {
        super(FeatureType.Issuer);
        this.address = issuer;
    }
}

/**
 * A Metadata feature.
 */
class MetadataFeature extends Feature {
    /** Defines metadata (arbitrary binary data) that will be stored in the output. */
    readonly entries: { [key: PrintableASCII]: HexEncodedString };

    /**
     * @param entries The metadata stored with the feature.
     */
    constructor(entries: { [key: PrintableASCII]: HexEncodedString }) {
        super(FeatureType.Metadata);
        this.entries = entries;
    }
}

/**
 * A Metadata Feature that can only be changed by the State Controller.
 */
class StateMetadataFeature extends Feature {
    /** Defines metadata (arbitrary binary data) that will be stored in the output. */
    readonly entries: { [key: PrintableASCII]: HexEncodedString };

    /**
     * @param entries The metadata stored with the feature.
     */
    constructor(entries: { [key: PrintableASCII]: HexEncodedString }) {
        super(FeatureType.StateMetadata);
        this.entries = entries;
    }
}

/**
 * A Tag feature.
 */
class TagFeature extends Feature {
    /** Defines a tag for the data. */
    readonly tag: string;

    /**
     * @param tag The tag stored with the feature.
     */
    constructor(tag: string) {
        super(FeatureType.Tag);
        this.tag = tag;
    }
}

/**
 * Native token feature.
 */
class NativeTokenFeature extends Feature {
    /**
     * Identifier of the native token.
     */
    readonly id: HexEncodedString;
    /**
     * Amount of native tokens of the given Token ID.
     */
    @Transform((value) => hexToBigInt(value.value))
    readonly amount: u256;

    /**
     * Creates a new `NativeTokenFeature`.
     * @param id The identifier of the native token.
     * @param amount The native token amount.
     */
    constructor(id: HexEncodedString, amount: u256) {
        super(FeatureType.NativeToken);
        this.id = id;
        this.amount = amount;
    }

    /**
     * Returns the native token contained in this feature.
     */
    public asNativeToken(): NativeToken {
        return {
            id: this.id,
            amount: this.amount,
        };
    }
}

/**
 * Block Issuer feature.
 */
class BlockIssuerFeature extends Feature {
    /**
     * The slot index at which the Block Issuer Feature expires and can be removed.
     */
    readonly expirySlot: SlotIndex;
    /**
     * The Block Issuer Keys.
     */
    @Type(() => BlockIssuerKey, {
        discriminator: BlockIssuerKeyDiscriminator,
    })
    readonly blockIssuerKeys: BlockIssuerKey[];

    constructor(expirySlot: SlotIndex, blockIssuerKeys: BlockIssuerKey[]) {
        super(FeatureType.BlockIssuer);
        this.expirySlot = expirySlot;
        this.blockIssuerKeys = blockIssuerKeys;
    }
}

/**
 * Staking feature.
 */
class StakingFeature extends Feature {
    /**
     * The amount of coins that are locked and staked in the containing account.
     **/
    readonly stakedAmount: u64;
    /**
     * The fixed cost of the validator, which it receives as part of its Mana rewards.
     */
    readonly fixedCost: u64;
    /**
     * The epoch index in which the staking started.
     */
    readonly startEpoch: EpochIndex;
    /**
     * The epoch index in which the staking ends.
     */
    readonly endEpoch: EpochIndex;

    constructor(
        stakedAmount: u64,
        fixedCost: u64,
        startEpoch: EpochIndex,
        endEpoch: EpochIndex,
    ) {
        super(FeatureType.Staking);
        this.stakedAmount = stakedAmount;
        this.fixedCost = fixedCost;
        this.startEpoch = startEpoch;
        this.endEpoch = endEpoch;
    }
}

const FeatureDiscriminator = {
    property: 'type',
    subTypes: [
        { value: SenderFeature, name: FeatureType.Sender as any },
        { value: IssuerFeature, name: FeatureType.Issuer as any },
        { value: MetadataFeature, name: FeatureType.Metadata as any },
        { value: StateMetadataFeature, name: FeatureType.StateMetadata as any },
        { value: TagFeature, name: FeatureType.Tag as any },
        { value: NativeTokenFeature, name: FeatureType.NativeToken as any },
        { value: BlockIssuerFeature, name: FeatureType.BlockIssuer as any },
        { value: StakingFeature, name: FeatureType.Staking as any },
    ],
};

export {
    FeatureDiscriminator,
    Feature,
    FeatureType,
    SenderFeature,
    IssuerFeature,
    MetadataFeature,
    StateMetadataFeature,
    TagFeature,
    NativeTokenFeature,
    BlockIssuerFeature,
    StakingFeature,
};
